flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java

```diff
@@ -83,9 +88,25 @@
 
 	// ------------------------------------------------------------------------
 
-	public AbstractMetricGroup(MetricRegistry registry, String[] scope) {
+	public AbstractMetricGroup(MetricRegistry registry, String[] scope, A parent) {
 		this.registry = checkNotNull(registry);
 		this.scopeComponents = checkNotNull(scope);
+		this.parent = parent;
+	}
+
+	public Map<String, String> getAllVariables() {
+		if (variables == null) { // avoid synchronization for common case
```

Comment:
This "double check locking" does not work properly, unless you make the variable `variables` volatile.
I am wondering if we should not simply initialize the map eagerly.